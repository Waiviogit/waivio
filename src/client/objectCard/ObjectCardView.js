import React, { useEffect, useState } from 'react';
import { useSelector } from 'react-redux';
import { filter, includes, orderBy, isEmpty } from 'lodash';
import { injectIntl } from 'react-intl';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import RatingsWrap from './RatingsWrap/RatingsWrap';
import WeightTag from '../components/WeightTag';
import DEFAULTS from '../object/const/defaultValues';
import { getFieldWithMaxWeight } from '../object/wObjectHelper';
import { getAuthenticatedUserName, getScreenSize } from '../reducers';
import { objectFields as objectTypes } from '../../common/constants/listOfFields';
import './ObjectCardView.less';

const ObjectCardView = ({
  intl,
  wObject,
  passedParent,
  options: { mobileView = 'compact', ownRatesOnly = false, pathNameAvatar = '' },
}) => {
  const screenSize = useSelector(getScreenSize);
  const username = useSelector(getAuthenticatedUserName);
  const [tags, setTags] = useState([]);

  useEffect(() => {
    if (wObject.tagCategories && wObject.tagCategories.length) {
      const currentTags = wObject.tagCategories
        .map(category => category.categoryItems)
        .filter(categoryItems => !!categoryItems.length)
        .map(items => orderBy(items, ['weight', 'name'])[0].name);
      setTags(currentTags);
    } else setTags([wObject.object_type]);
  }, []);

  const getObjectRatings = () => {
    const ratingFields = filter(wObject.fields, ['name', 'rating']);
    return ownRatesOnly
      ? ratingFields.map(rating => ({
          ...rating,
          rating_votes:
            (rating.rating_votes && rating.rating_votes.filter(vote => vote.voter === username)) ||
            [],
        }))
      : ratingFields;
  };
  const pathName = pathNameAvatar || `/object/${wObject.id}`;
  const ratings = getObjectRatings();

  const avatarLayout = (avatar = DEFAULTS.AVATAR) => {
    let url = avatar;
    if (!isEmpty(passedParent) && avatar === DEFAULTS.AVATAR) {
      url = passedParent.avatar;
    }
    if (includes(url, 'waivio.')) url = `${url}_medium`;

    return (
      <div
        className="avatar-image"
        style={{
          backgroundImage: `url(${url})`,
          backgroundSize: 'cover',
          backgroundPosition: 'center',
        }}
      />
    );
  };
  const objName = wObject.name || wObject.default_name;
  const parentName = isEmpty(passedParent)
    ? getFieldWithMaxWeight(wObject.parent, objectTypes.name, '')
    : passedParent.name || passedParent.default_name;

  const goToObjTitle = wobjName =>
    `${intl.formatMessage({
      id: 'GoTo',
      defaultMessage: 'Go to',
    })} ${wobjName}`;
  return (
    <React.Fragment>
      <div className="ObjectCardView">
        <div className="ObjectCardView__content">
          <div className="ObjectCardView__content-row">
            <Link to={pathName} title={goToObjTitle(objName)} className="ObjectCardView__avatar">
              {avatarLayout(wObject.avatar)}
            </Link>
            <div className="ObjectCardView__info">
              {parentName && (
                <Link
                  to={`/object/${
                    isEmpty(passedParent)
                      ? wObject.parent.author_permlink
                      : passedParent.author_permlink
                  }`}
                  title={goToObjTitle(parentName)}
                  className="ObjectCardView__type"
                >
                  {parentName}
                </Link>
              )}
              <div className="ObjectCardView__name">
                <Link
                  to={pathName}
                  className="ObjectCardView__name-truncated"
                  title={goToObjTitle(objName)}
                >
                  {objName}
                </Link>
                {!isNaN(wObject.weight) && <WeightTag weight={Number(wObject.weight)} />}
              </div>
              {ratings && (
                <RatingsWrap
                  mobileView={mobileView}
                  ownRatesOnly={ownRatesOnly}
                  ratings={ratings}
                  screenSize={screenSize}
                  username={username}
                  wobjId={wObject.id || wObject.author_permlink}
                  wobjName={wObject.name || wObject.default_name}
                />
              )}
              <span className="ObjectCardView__tag-text">
                {wObject.price && (
                  <span className="ObjectCardView__price" title={wObject.price}>
                    {wObject.price}
                  </span>
                )}
                {tags.map((tag, index) => (
                  <span key={tag}>
                    {index === 0 && !wObject.price ? tag : <span>&nbsp;&middot;{` ${tag}`}</span>}
                  </span>
                ))}
              </span>
              {wObject.address && (
                <div className="ObjectCardView__tag-text">
                  {wObject.address.street && `${wObject.address.street}, `}
                  {wObject.address.city}
                </div>
              )}
              {/* eslint-disable-next-line no-nested-ternary */}
              {wObject.title ? (
                <div className="ObjectCardView__title" title={wObject.title}>
                  {wObject.title}
                </div>
              ) : wObject.description ? (
                <div className="ObjectCardView__title" title={wObject.description}>
                  {wObject.description}
                </div>
              ) : null}
            </div>
          </div>
        </div>
      </div>
    </React.Fragment>
  );
};

ObjectCardView.propTypes = {
  intl: PropTypes.shape().isRequired,
  wObject: PropTypes.shape().isRequired,
  options: PropTypes.shape({
    mobileView: PropTypes.oneOf(['compact', 'full']),
    ownRatesOnly: PropTypes.bool,
    pathNameAvatar: PropTypes.oneOfType([PropTypes.string, PropTypes.shape()]),
  }),
  passedParent: PropTypes.shape(),
};

ObjectCardView.defaultProps = {
  options: {},
  passedParent: {},
};
export default injectIntl(ObjectCardView);
