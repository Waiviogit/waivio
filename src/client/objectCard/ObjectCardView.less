@import (reference) '../styles/custom.less';
@import (reference) '../styles/modules/variables.less';

@wrapPadding: 12px;

.ObjectCardView {
  font-size: 16px;
  display: flex;
  justify-content: space-between;
  border: 1px solid @blue-hawkesBlue;
  border-radius: 4px;
  margin-top: 5px;
  &__content {
    width: 100%;
    position: relative;
    display: flex;
    flex-flow: column nowrap;
    min-height: 80px;
    outline: none;
    margin: @wrapPadding;
  }
  &__content-row {
    display: flex;
    flex-flow: row nowrap;
    width: 100%;
  }
  &__info {
    width: 70%;
    max-width: 70%;
    padding-left: @wrapPadding;
    word-wrap: break-word;
    word-break: break-word;
    white-space: normal;
    &.small {
      .ObjectCardView__title,
      .ObjectCardView__name-truncated {
        font-size: 1em;
      }
    }
  }
  &__avatar {
    width: 25%;
    .avatar-image {
      width: 100%;
      padding-top: 100%;
      border-radius: 8px;
      display: block !important;
      background-size: cover;
      background-repeat: no-repeat;
      background-position: 50% 50%;
      border: 1px solid #e9e7e7;
    }
  }
  &__type {
    color: @blue-botticelli;
    margin-bottom: 0.5em;
  }
  &__price {
    color: @blue-botticelli;
    margin-bottom: 0.5em;
    i {
      margin-right: 5px;
    }
  }
  &__name {
    cursor: auto;
    display: flex;
    align-items: baseline;

    &-truncated {
      margin-right: 10px;
      color: @black;
      font-size: 1.2em;
      line-height: 1.1em;
      white-space: normal;
      vertical-align: middle;
      font-weight: 600;
    }
  }
  &__title {
    margin-top: 0.5em;
    color: @black;
    font-size: 1.2em;
  }
}

@import 'ObjectCardView-nightmode';
