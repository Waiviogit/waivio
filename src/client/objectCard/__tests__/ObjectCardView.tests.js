import React from 'react';
import { shallow } from 'enzyme';
import { IntlProvider } from 'react-intl';
import ObjectCardView from '../ObjectCardView';

describe('(Component) ObjectCardViewwith default props', () => {
  it('renders and matches snapshot', () => {
    const wrapper = shallow(
      <IntlProvider>
        <ObjectCardView />
      </IntlProvider>,
    );
    expect(wrapper).toMatchSnapshot();
  });
});
