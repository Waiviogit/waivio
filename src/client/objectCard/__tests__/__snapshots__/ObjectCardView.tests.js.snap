// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`(Component) ObjectCardViewwith default props renders and matches snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <IntlProvider
    locale="en"
  >
    <InjectIntl(ObjectCardView) />
  </IntlProvider>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {},
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {},
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    Symbol(enzyme.__providerValues__): undefined,
  },
  Symbol(enzyme.__providerValues__): Map {},
  Symbol(enzyme.__childContext__): Object {
    "intl": Object {
      "defaultFormats": Object {},
      "defaultLocale": "en",
      "formatDate": [Function],
      "formatHTMLMessage": [Function],
      "formatMessage": [Function],
      "formatNumber": [Function],
      "formatPlural": [Function],
      "formatRelative": [Function],
      "formatTime": [Function],
      "formats": Object {},
      "formatters": Object {
        "getDateTimeFormat": [Function],
        "getMessageFormat": [Function],
        "getNumberFormat": [Function],
        "getPluralFormat": [Function],
        "getRelativeFormat": [Function],
      },
      "locale": "en",
      "messages": Object {},
      "now": [Function],
      "onError": [Function],
      "textComponent": "span",
      "timeZone": null,
    },
  },
}
`;
